File: runtime/adapters/openai_provider_adapter.py
"""
OpenAI Provider Adapter - Subordinate to Inference Authority

Constitutional Law: This is a PROVIDER ADAPTER, not an authority.
It is subordinate to InferenceAdapter.

This adapter provides OpenAI-compatible implementation (vLLM).
Only InferenceAdapter may call this adapter.
No business logic may call this adapter directly.
"""

import requests
from typing import List, Dict, Any, Optional

class OpenAIProviderAdapter:
    """OpenAI-compatible provider adapter (subordinate to InferenceAdapter)."""
    
    def __init__(
        self,
        base_url: str,
        embedding_model: str,
        chat_model: str
    ):
        self.base_url = base_url
        self.embedding_model = embedding_model
        self.chat_model = chat_model
    
    def embed(self, text: str) -> Optional[List[float]]:
        """Generate embedding via OpenAI-compatible endpoint."""
        try:
            response = requests.post(
                f"{self.base_url}/v1/embeddings",
                json={
                    "model": self.embedding_model,
                    "input": text
                },
                timeout=60
            )
            
            if response.status_code == 200:
                data = response.json()
                embeddings = data.get('data', [])
                if embeddings and len(embeddings) > 0:
                    return embeddings[0].get('embedding')
                else:
                    print("No embedding in response")
                    return None
            else:
                print(f"Embedding failed: {response.status_code}")
                return None
        except Exception as e:
            print(f"Embedding error: {e}")
            return None
    
    def chat(
        self,
        messages: List[Dict[str, str]],
        options: Optional[Dict[str, Any]] = None
    ) -> Optional[Dict[str, Any]]:
        """Generate chat completion via OpenAI-compatible endpoint."""
        try:
            payload = {
                "model": self.chat_model,
                "messages": messages
            }
            
            if options:
                payload.update(options)
            
            response = requests.post(
                f"{self.base_url}/v1/chat/completions",
                json=payload,
                timeout=120
            )
            
            if response.status_code == 200:
                data = response.json()
                # Convert OpenAI format to Ollama format for compatibility
                return {
                    "message": {
                        "role": "assistant",
                        "content": data['choices'][0]['message']['content']
                    },
                    "model": data['model'],
                    "done": True
                }
            else:
                print(f"Chat failed: {response.status_code}")
                return None
        except Exception as e:
            print(f"Chat error: {e}")
            return None
    
    def health(self) -> bool:
        """Check if provider is healthy."""
        try:
            response = requests.get(f"{self.base_url}/v1/models", timeout=10)
            return response.status_code == 200
        except Exception as e:
            print(f"Health check failed: {e}")
            return False
    
    def model_capability(self, model: str, capability: str) -> bool:
        """Check if model has specific capability."""
        try:
            response = requests.get(f"{self.base_url}/v1/models", timeout=10)
            if response.status_code == 200:
                models = response.json().get('data', [])
                for m in models:
                    if m['id'] == model:
                        return True
                return False
            return False
        except Exception as e:
            print(f"Capability check failed: {e}")
            return False
    
    def list_models(self) -> Optional[List[Dict[str, Any]]]:
        """List available models."""
        try:
            response = requests.get(f"{self.base_url}/v1/models")
            if response.status_code == 200:
                data = response.json()
                return data.get('data', [])
            return None
        except Exception as e:
            print(f"List models failed: {e}")
            return None
    
    def get_model_metadata(self, model: str) -> Optional[Dict[str, Any]]:
        """Get metadata for a specific model."""
        try:
            # OpenAI-compatible APIs don't have a standard /show endpoint
            # Return basic model info from list
            models = self.list_models()
            if models:
                for m in models:
                    if m.get('id') == model:
                        return m
            return None
        except Exception as e:
            print(f"Get model metadata failed: {e}")
            return None
    
    def pull_model(self, model: str) -> bool:
        """Pull model - not supported for OpenAI-compatible providers."""
        raise NotImplementedError("Model pulling not supported for OpenAI-compatible providers")

File: gateway/openai_provider_adapter.js
/**
 * OpenAI Provider Adapter - Subordinate to Inference Authority (Node.js)
 * 
 * Constitutional Law: This is a PROVIDER ADAPTER, not an authority.
 * It is subordinate to InferenceAdapter.
 * 
 * This adapter provides OpenAI-compatible implementation (vLLM).
 * Only InferenceAdapter may call this adapter.
 * No business logic may call this adapter directly.
 */

class OpenAIProviderAdapter {
  constructor(baseUrl, embeddingModel, chatModel) {
    this.baseUrl = baseUrl;
    this.embeddingModel = embeddingModel;
    this.chatModel = chatModel;
  }
  
  async embed(text) {
    try {
      const response = await fetch(`${this.baseUrl}/v1/embeddings`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: this.embeddingModel,
          input: text
        })
      });
      
      if (response.ok) {
        const data = await response.json();
        const embeddings = data.data || [];
        if (embeddings.length > 0) {
          return embeddings[0].embedding;
        } else {
          console.error('No embedding in response');
          return null;
        }
      } else {
        console.error(`Embedding failed: ${response.status}`);
        return null;
      }
    } catch (error) {
      console.error(`Embedding error: ${error.message}`);
      return null;
    }
  }
  
  async chat(messages, options = null) {
    try {
      const payload = {
        model: this.chatModel,
        messages: messages
      };
      
      if (options) {
        Object.assign(payload, options);
      }
      
      const response = await fetch(`${this.baseUrl}/v1/chat/completions`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
      });
      
      if (response.ok) {
        const data = await response.json();
        // Convert OpenAI format to Ollama format for compatibility
        return {
          message: {
            role: 'assistant',
            content: data.choices[0].message.content
          },
          model: data.model,
          done: true
        };
      } else {
        console.error(`Chat failed: ${response.status}`);
        return null;
      }
    } catch (error) {
      console.error(`Chat error: ${error.message}`);
      return null;
    }
  }
  
  async health() {
    try {
      const response = await fetch(`${this.baseUrl}/v1/models`);
      return response.ok;
    } catch (error) {
      console.error(`Health check failed: ${error.message}`);
      return false;
    }
  }
  
  async modelCapability(model, capability) {
    try {
      const response = await fetch(`${this.baseUrl}/v1/models`);
      if (response.ok) {
        const data = await response.json();
        const models = data.data || [];
        return models.some(m => m.id === model);
      }
      return false;
    } catch (error) {
      console.error(`Capability check failed: ${error.message}`);
      return false;
    }
  }
}

module.exports = { OpenAIProviderAdapter };
